instant.ts

```diff
@@ -74,6 +74,89 @@ async function runTests(path: string) {
   }
 }
 
+const orderPackageIds = (allPackages, chosenPackageIds) => {
+  const orderedPackageIds = []
+  const packagesWithDependencies = []
+
+  function collectPackagesWithDependencies(chosenIds) {
+    let inexplicitDependencies = []
+
+    chosenIds.forEach(id => {
+      if (id && allPackages[id] && allPackages[id].metadata) {
+        if (
+          !allPackages[id].metadata.dependencies ||
+          !allPackages[id].metadata.dependencies.length
+        ) {
+          if (id === 'core') {
+            orderedPackageIds.unshift(id)
+          } else {
+            orderedPackageIds.push(id)
+          }
+        } else {
+          allPackages[id].metadata.dependencies.forEach(dependency => {
+            if (
+              !Object.keys(allPackages).includes(dependency) ||
+              !allPackages[dependency].metadata
+            ) throw Error(
+              `Dependency ${dependency} for package ${id} does not exist or the metadata is invalid`
+            )
+            if (
+              allPackages[dependency].metadata.dependencies &&
+              allPackages[dependency].metadata.dependencies.length &&
+              allPackages[dependency].metadata.dependencies.includes(id)
```

Comment:
Thinking about this recursive function. I wonder if you could expand the inexplict dependencies and do the ordering like in the `while` loop below in a single recursive function as you are already having to go through every dependency. This might simplify this alot. Something like:

```
resolveDeps(id, currentStack) {
  if currentStack contains id
    throw circular dep error
  else push id to currentStack

  if id has no deps
    return [id]

  ordferedIds = []
  for each dep of id
    orderedIds = resolveDeps(dep, currentStack)
    orderedIds.push(dep)

  return orderedIds
}

finalOrderedPackages = []
foreach chosenId
  finalOrderedPackages.push(resolveDeps(chosenId, []))

// at this point you may have to remove any duplicate package Ids from the array.
```

Reply:
good solution you have here. I've addressed all in https://github.com/openhie/instant/pull/63/commits/f983bc037e991eb4713fa96a35e56d1e8e680e35